core/coreunix/reed_solomon_add.go

```diff
@@ -0,0 +1,211 @@
+package coreunix
+
+import (
+	"errors"
+	"io"
+
+	"encoding/json"
+	"github.com/TRON-US/go-btfs-files"
+	"github.com/TRON-US/go-unixfs"
+	uio "github.com/TRON-US/go-unixfs/io"
+	ipld "github.com/ipfs/go-ipld-format"
+	dag "github.com/ipfs/go-merkledag"
+	gopath "path"
+)
+
+type ReedSolomonAdder struct {
+	Adder
+	InfileReaders []io.Reader
+	IsDir         bool
+	CurrentOffset uint64
+}
+
+// NewReedSolomonAdder returns a new ReedSolomonAdder used for a file/directory add operation.
+func NewReedSolomonAdder(adder *Adder) (*ReedSolomonAdder, error) {
+	rsa := &ReedSolomonAdder{
+		InfileReaders: nil,
+		IsDir:         false,
+		CurrentOffset: 0,
+	}
+	rsa.Adder = *adder
+
+	return rsa, nil
+}
+
+// AddAllAndPin adds the given request's files and pin them.
+func (rsadder *ReedSolomonAdder) AddAllAndPin(file files.Node) (ipld.Node, error) {
+	if rsadder.Pin {
+		rsadder.unlocker = rsadder.gcLocker.PinLock()
+	}
+	defer func() {
+		if rsadder.unlocker != nil {
+			rsadder.unlocker.Unlock()
+		}
+	}()
+
+	// Create a directory tree from the input `file` directory.
+	n, err := rsadder.addFileNode("", file, true)
+	if err != nil {
+		return nil, err
+	}
+	var byts []byte
+	switch node := n.(type) {
+	case *uio.DirNode, *uio.SymlinkNode, *uio.FileNode:
+		byts, err = json.Marshal(node)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, errors.New("unexpected node type")
+	}
+
+	var reader io.Reader = io.MultiReader(rsadder.InfileReaders...)
+	if rsadder.Progress {
+		reader = &progressReader{file: reader, path: "", out: rsadder.Out}
+	}
+
+	// Create a DAG with the above directory tree as metadata and
+	// the data from the given `file` directory or file.
+	nd, err := rsadder.add(reader, byts)
+	if err != nil {
+		return nil, err
+	}
+
+	// output directory and file events
+	err = rsadder.outputDirs("", nd)
+	if err != nil {
+		return nil, err
+	}
+
+	// Pin the newly created DAG.
+	if !rsadder.Pin {
+		return nd, nil
+	}
+
+	if err := rsadder.PinRoot(nd); err != nil {
+		return nil, err
+	}
+
+	return nd, nil
+}
+
+// addFileNode traverses the directory tree under
+// the given `file` in a bottom up DFS way.
+func (rsadder *ReedSolomonAdder) addFileNode(path string, file files.Node, toplevel bool) (uio.Node, error) {
+	//defer file.Close()
```

Comment:
can you add comment why file won't be closed?

Reply:
Modified the code to close after add(). Thank you.